vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/resource/RestApi.java

```diff
@@ -278,27 +281,81 @@ String getParameter() {
+            // to enforce this symmetry when trying to retrieve them as well.
+            Text.validateTextString(group.value).ifPresent(codepoint -> {
+                throw new BadRequestParameterException(SELECTION, String.format(
+                        "Failed to parse group part of selection URI; contains invalid text code point U%04X", codepoint));
+            });
+            return String.format("id.group=='%s'", singleQuoteEscapedString(group.value));
+        }
+    }
+
+    private static void validateDocumentSelectionSyntax(String expression) {
+        try {
+            new DocumentSelector(expression);
+        } catch (ParseException e) {
+            throw new BadRequestParameterException(SELECTION, String.format("Failed to parse expression given in 'selection'" +
+                    " parameter. Must be a complete and valid sub-expression. Error: %s", e.getMessage()));
+        }
+    }
+
+    private static String documentSelectionFromRequest(RestUri restUri, HttpRequest request) throws BadRequestParameterException {
         String documentSelection = Optional.ofNullable(request.getProperty(SELECTION)).orElse("");
+        if (!documentSelection.isEmpty()) {
+            // Ensure that the selection parameter sub-expression is complete and valid by itself.
+            validateDocumentSelectionSyntax(documentSelection);
+        }
         if (restUri.getGroup().isPresent() && ! restUri.getGroup().get().value.isEmpty()) {
-            if (! documentSelection.isEmpty()) {
-                // TODO why is this restriction in place? Document selection allows composition of location predicate and other expressions
-                return Response.createErrorResponse(
-                        400,
-                        "Visiting does not support setting value for group/value in combination with expression, try using only expression parameter instead.",
-                        restUri,
-                        RestUri.apiErrorCodes.GROUP_AND_EXPRESSION_ERROR);
-            }
-            RestUri.Group group = restUri.getGroup().get();
-            if (group.name == 'n') {
-                documentSelection = "id.user=" + group.value;
+            String locationSubExpression = validateAndBuildLocationSubExpression(restUri.getGroup().get());
+            if (documentSelection.isEmpty()) {
+                documentSelection = locationSubExpression;
             } else {
-                documentSelection = "id.group='" + group.value + "'";
+                documentSelection = String.format("%s and (%s)", locationSubExpression, documentSelection);
             }
         }
+        return documentSelection;
+    }
+
+    private HttpResponse handleVisit(RestUri restUri, HttpRequest request) throws RestApiException {
+        String documentSelection;
         OperationHandler.VisitOptions options;
         try {
+            documentSelection = documentSelectionFromRequest(restUri, request);
```

Comment:
Why was there 1 `=`  in `documentSelection` before, but 2 `=` now?

Reply:
The value of the (old) `documentSelection` previously had one equals sign between, for example, `id.group` and its value (https://github.com/vespa-engine/vespa/blob/177392cd29ab9db2de0471d295cce0d0423ed726/vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/resource/RestApi.java#L297)

But now there are two equal signs (https://github.com/vespa-engine/vespa/blob/932812ddcd333ce57c0a5d812b12d4f0ced971a5/vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/resource/RestApi.java#L324)

Reply:
Ahh yes, sorry 🙂 Basically, the `=` operator in document selection is a multi-purpose operator which for integers degenerates to `==`, but for strings can end up as glob-style operator that uses regexes behind the scenes. Operator `==` always implies a concrete match, which is the semantics we want here.

I changed this to ensure we have a deterministic performance profile for comparisons. No reason to use `=` here in the first place, though it _does_ functionally work as intended.